viv_utils/emulator_drivers.py

```diff
@@ -178,131 +226,226 @@ def doHook(self, pc, op):
 
         for mon in self._monitors:
             try:
-                r = mon.apicall(self, op, pc, api, argv)
-                if r is not None:
-                    # take the first result
-                    # not ideal, but works in the common case
-                    self._logger.debug("monitor hook handled call: %s", callname)
-                    return True
+                r = mon.apicall(self, api, argv)
             except StopEmulation:
                 raise
             except Exception as e:
-                mon.logAnomaly(emu, pc,
-                        "%s.apicall failed: %s" % (mon.__class__.__name__, e))
+                logger.debug("driver: %s.apicall failed: %s", mon.__class__.__name__, e)
+                continue
+            else:
+                if r:
+                    # note: short circuit
+                    logger.debug("driver: %s.apicall: handled call: %s", mon.__class__.__name__, callname)
+                    return True
 
         for hook in self._hooks:
             try:
-                ret = hook.hook(callname, self, callconv, api, argv)
-                # take the first result
-                # not ideal, but works in the common case
-                if ret is not None:
-                    self._logger.debug("driver hook handled call: %s", callname)
-                    return True
+                ret = hook(self, api, argv)
             except StopEmulation:
                 raise
-            except UnsupportedFunction:
-                continue
             except Exception as e:
-                mon.logAnomaly(emu, pc,
-                        "%s.apicall failed: %s" % (hook.__class__.__name__, e))
-        if callname and callname not in ("UnknownApi", "?"):
-            self._logger.debug("driver hook API call NOT handled: %s", callname)
+                logger.debug("driver: hook: %r failed: %s", hook, e)
+                continue
+            else:
+                if ret:
+                    # note: short circuit
+                    logger.debug("driver: hook handled call: %s", callname)
+                    return True
 
         if callname in emu.hooks:
+            # this is where vivisect-internal hooks are stored,
+            # such as those provided by impapi.
+            # note that we prefer locally configured hooks, first.
             hook = emu.hooks.get(callname)
             try:
-                hook(self, callconv, api, argv)
-                self._logger.debug("emu hook handled call: %s", callname)
-                return True
+                hook(self, api, argv)
```

Comment:
vivisect defines hook with a different API, e.g. `def kernel32_GetModuleHandleA(self, emu, callconv, api, argv)`

should we keep the hook API consistent or pass the modified arguments here?